miqa/core/rest/scan_decision.py

```diff
@@ -23,23 +24,30 @@ class Meta:
 class ScanDecisionViewSet(
     mixins.CreateModelMixin,
     mixins.RetrieveModelMixin,
-    mixins.DestroyModelMixin,
     mixins.ListModelMixin,
     GenericViewSet,
 ):
-    queryset = ScanDecision.objects.select_related('scan__experiment__project')
-
     filter_backends = [filters.DjangoFilterBackend]
-    filterset_fields = ['scan', 'creator']
-
     permission_classes = [IsAuthenticated, UserHoldsExperimentLock]
-
     serializer_class = ScanDecisionSerializer
 
-    def create(self, request, *args, **kwargs):
+    def get_queryset(self):
+        projects = get_objects_for_user(
+            self.request.user,
+            [f'core.{perm}' for perm in Project.get_read_permission_groups()],
+            any_perm=True,
+        )
+        return ScanDecision.objects.filter(scan__experiment__project__in=projects)
+
+    # cannot use project_permission_required decorator because no pk is provided
+    def create(self, request, **kwargs):
         request_data = request.data
-        request_data['scan'] = Scan.objects.get(id=request.data['scan'])
+        scan = Scan.objects.get(id=request.data['scan'])
+        if get_perms(request.user, scan.experiment.project) == []:
+            return Response(status=status.HTTP_401_UNAUTHORIZED)
```

Comment:
This looks like it implicitly checks if the user has reviewer access or higher, but that's not really clear from the code at all, and seems rather fragile. It might be worth breaking the decorator into two parts: a helper `has_project_permission(user, project, review_access=False, superuser_access=False)`, and the decorator which calls it. Then you can call `has_project_permission` here or anywhere else you might want to do an explicit permission check.

Reply:
Also, breaking it up allows you to write a unit test for the `has_project_permission` function in isolation, which is generally a good idea when you have a big chunk of code like that. As it is I'm not sure how you would go about writing a test for `@project_permission_required` in isolation, although we do implicitly test it by testing the views that use it.

Reply:
Good point. I wan't able to turn it into a decorator because the pk of a db object is not passed through the parameters of this PUT request. However, I made functions shared by the decorator itself and the endpoint function. These functions have a more concrete way of testing for the proper permissions.